keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,394 @@
+      - [Resources vs. Patches](#resources-vs-patches)
+      - [Task Application](#task-application)
+    - [Parameters](#parameters)
+    - [Templates](#templates)
+  - [Extensions and Bases](#extensions-and-bases)
+    - [Task Extensions](#task-extensions)
+    - [Plan Extensions](#plan-extensions)
+    - [Example Framework Extension](#example-framework-extension)
+      - [framework.yaml](#frameworkyaml-1)
+      - [params.yaml](#paramsyaml-1)
+    - [KUDO](#kudo)
+    - [Helm](#helm)
+  - [Execution State](#execution-state)
+  - [Example Framework](#example-framework)
+  - [Future Work](#future-work)
+    - [Allow for other templating engines](#allow-for-other-templating-engines)
+
+## Summary
+
+Drive KUDO to have a set of tools and specifications for creating frameworks.
+
+## Motivation
+
+KUDO provides a way to reduce the amount of code necessary for the creation of an framework for Kubernetes. The current implementation of KUDO requires a significant amount of YAML to define an framework, located in one file. This YAML relies in nested YAML inlined via multiline strings, is not extensible, and is prone to error. This KEP defines an framework SDK that solves these issues.
+
+## Goals
+
+The toolkit must provide the ability for:
+
+- conditional inclusions
+- defining default values
+- replacement of values
+- removal of keys and values
+
+## Proposal
+
+A KUDO framework is split into several base components, ordered from the outside-in:
+
+- [Plans](#plans)
+- [Steps](#steps)
+- [Tasks](#tasks)
+- [Templates](#templates)
+- [Parameters](#parameters)
+- [Extensions and Bases](#extensions-and-bases)
+- [Execution State](#execution-state)
+
+These combine to form an framework responsible for the deployment, upgrading, and day 2 operations of software deployed on Kubernetes. While KUDO ships with a templating system, it is not intended to advance the state of deploying software on Kubernetes, and provides facilities for integrating with other templating and deployment systems. KUDO's focus is instead on the sequencing and day 2 operations of this software, and is able to take advantage of existing templating to fulfill this need.
+
+### Definitions
+
+### Framework Organization
+
+An framework bundle is a folder that contains all of the manifests needed to create or extend a KUDO framework. In the most basic form, an framework bundle is structured in the following format:
+
+```shell
+.
+├── framework.yaml
+├── params.yaml
+└── common
+    └── common.yaml
+└── templates
+    ├── backup.yaml
+    ├── clear-data.yaml
+    ├── deployment.yaml
+    ├── init.yaml
+    ├── load-data.yaml
+    ├── pvc.yaml
+    ├── restore.yaml
+    └── service.yaml
+```
+
+#### framework.yaml
+
+`framework.yaml` is the base definition of an framework. It follows the following format, extracted from the MySQL example:
+
+```yaml
+name: "my-framework"
+description: A test example framework.
+version: "5.7"
+kudoVersion: ">= 0.2.0"
+kubeVersion: ">= 1.14"
```

Comment:
Can we be more explicit with this name ? `kubeVersion` could be misunderstood as the name for the running version on the kubelet which could sometimes differ from the version of the control-plane. I would suggest to rename it to `kubernetesVersion` instead.

Reply:
Yes. Will alter to `kubernetesVersion`.